apps/lib/apps.c

```diff
@@ -551,71 +572,96 @@ X509_REQ *load_csr(const char *file, int format, const char *desc)
         print_format_error(format, OPT_FMT_PEMDER);
 
  end:
-    if (req == NULL && desc != NULL) {
+    if (req == NULL) {
         BIO_printf(bio_err, "Unable to load %s\n", desc);
         ERR_print_errors(bio_err);
     }
     BIO_free(in);
     return req;
 }
 
-EVP_PKEY *load_key(const char *file, int format, int maybe_stdin,
+void cleanse(char *str)
+{
+    if (str != NULL)
+        OPENSSL_cleanse(str, strlen(str));
+}
+
+void clear_free(char *str)
+{
+    if (str != NULL)
+        OPENSSL_clear_free(str, strlen(str));
+}
+
+EVP_PKEY *load_key(const char *uri, int format, int maybe_stdin,
                    const char *pass, ENGINE *e, const char *desc)
 {
-    BIO *key = NULL;
+    BIO *bio = NULL;
     EVP_PKEY *pkey = NULL;
     PW_CB_DATA cb_data;
 
+    if (desc == NULL)
+        desc = "private key";
     cb_data.password = pass;
-    cb_data.prompt_info = file;
+    cb_data.prompt_info = uri;
 
-    if (file == NULL && (!maybe_stdin || format == FORMAT_ENGINE)) {
-        BIO_printf(bio_err, "No keyfile specified\n");
-        goto end;
-    }
     if (format == FORMAT_ENGINE) {
         if (e == NULL) {
-            BIO_printf(bio_err, "No engine specified\n");
+            BIO_printf(bio_err, "No engine specified for loading %s\n", desc);
         } else {
 #ifndef OPENSSL_NO_ENGINE
             if (ENGINE_init(e)) {
-                pkey = ENGINE_load_private_key(e, file,
+                pkey = ENGINE_load_private_key(e, uri,
                                                (UI_METHOD *)get_ui_method(),
                                                &cb_data);
                 ENGINE_finish(e);
             }
-            if (pkey == NULL && desc != NULL) {
+            if (pkey == NULL) {
                 BIO_printf(bio_err, "Cannot load %s from engine\n", desc);
                 ERR_print_errors(bio_err);
             }
 #else
-            BIO_printf(bio_err, "Engines not supported\n");
+            BIO_printf(bio_err, "Engines not supported for loading %s\n", desc);
 #endif
         }
         goto end;
     }
-    if (file == NULL && maybe_stdin) {
+    if (uri == NULL) {
+        if (!maybe_stdin) {
+            BIO_printf(bio_err, "No filename or uri specified for loading %s\n",
+                       desc);
+            goto end;
+        }
         unbuffer(stdin);
-        key = dup_bio_in(format);
-    } else {
-        key = bio_open_default(file, 'r', format);
+        bio = dup_bio_in(format);
```

Comment:
It's possible to use the STORE API with stdin.  The function is calledd `OSSL_STORE_attach` and is to be used in place of `OSSL_STORE_open` when you want it to handle a BIO using some chosen scheme.  The OSSL_STORE functionality knows how to handle streamed data, it uses the cache buffer to catch the first 4KiB and uses that to find out (guess) the format.

Reply:
Hmm, would be nice, but `OSSL_STORE_attach()` does not exist, just undocumented internal functions `ossl_store_attach_pem_bio()` and `ossl_store_file_attach_pem_bio_int()`.

How if I export something like
```
OSSL_STORE_attach(bio, ui_method, ui_data) = 
ossl_store_attach_pem_bio(bio, ui_method, ui_data, NULL, NULL)
```
?

Reply:
Ah, right, it *could* be that it's because I looked at my own effort to do the same thing, where that is indeed already done, #7390

I should probably split that one out a bit.

Reply:
@levitte, I've found a much simpler solution than using `OSSL_STORE_attach()` (which also would require including part of #7390) - see the tiny commit 8b1080a in #11755.